server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java

```diff
@@ -30,139 +28,129 @@
-        List<String> unmappedConcreteFields = partitions.get(false);
-        if (unmappedWildcardPattern.isEmpty() == false) {
-            unmappedFieldsFetchAutomaton = new CharacterRunAutomaton(
-                Regex.simpleMatchToAutomaton(unmappedWildcardPattern.toArray(new String[unmappedWildcardPattern.size()])),
-                AUTOMATON_MAX_DETERMINIZED_STATES
-            );
+    private static ValueFetcher buildValueFetcher(SearchExecutionContext context, ResolvedField fieldAndFormat) {
+        try {
+            return fieldAndFormat.ft.valueFetcher(context, fieldAndFormat.format);
+        } catch (IllegalArgumentException e) {
+            StringBuilder error = new StringBuilder("error fetching [").append(fieldAndFormat.field).append(']');
+            if (fieldAndFormat.matchingPattern != null) {
+                error.append(" which matched [").append(fieldAndFormat.matchingPattern).append(']');
+            }
+            error.append(": ").append(e.getMessage());
+            throw new IllegalArgumentException(error.toString(), e);
         }
-        return new FieldFetcher(fieldContexts, unmappedFieldsFetchAutomaton, unmappedConcreteFields);
+    }
+
+    private static Map<String, FieldContext> buildFieldContexts(
+        SearchExecutionContext context,
+        String nestedScope,
+        List<ResolvedField> fields,
+        List<String> unmappedFetchPatterns
+    ) {
+
+        final boolean includeUnmapped = unmappedFetchPatterns.isEmpty() == false;
+
+        List<String> nestedMappers = context.nestedLookup().getImmediateChildMappers(nestedScope);
+        Map<String, List<ResolvedField>> fieldsByNestedMapper = NestedUtils.partitionByChildren(
+            nestedScope,
+            nestedMappers,
+            fields,
+            f -> f.field
+        );
+
+        // Using a LinkedHashMap so fields are returned in the order requested.
+        // We won't formally guarantee this, but but it's good for readability of the response
+        Map<String, FieldContext> output = new LinkedHashMap<>();
+        for (String scope : fieldsByNestedMapper.keySet()) {
+            if (nestedScope.equals(scope)) {
+                // These are fields in the current scope, so add them directly to the output map
+                for (ResolvedField ff : fieldsByNestedMapper.get(nestedScope)) {
+                    output.put(ff.field, new FieldContext(ff.field, buildValueFetcher(context, ff)));
+                }
+            } else {
+                // don't create nested fetchers if no children have been requested as part of the fields
+                // request, unless we are trying to also fetch unmapped fields``
+                if (includeUnmapped || fieldsByNestedMapper.get(scope).isEmpty() == false) {
+                    // These fields are in a child scope, so build a nested mapper for them
+                    Map<String, FieldContext> scopedFields = buildFieldContexts(
```

Comment:
so this is the recursion step here. Am I reading this right in that the partitioning step above (NestedUtils.partitionByChildren) is whats finally reducing the input and will finally stop the recursion?